File: 30days-of-code/28_regex.py
#!/bin/python3

import re


def read_input():
    total = int(input())

    items = []
    for _ in range(total):
        items.append(input().split())
    return items


def has_gmail(emails):
    ''' filter is_gmail emails '''
    gmail_addr = []
    gmail_pattern = re.compile(r'[a-z]+(\.[a-z]+)?@gmail\.com$')

    for (first_name, email_addr) in emails:
        is_gmail = gmail_pattern.match(email_addr)
        if is_gmail:
            gmail_addr.append(first_name)

    return gmail_addr


if __name__ == '__main__':
    # read | filter | show
    emails_in = read_input()

    for name in sorted(has_gmail(emails_in)):
        print(name)

File: 30days-of-code/28_regex.cc
#include <bits/stdc++.h>

using namespace std;

using Emails = vector<pair<string, string>>;


template <typename T>
pair<T,T> readPair() {
    T fst{}, snd{};
    cin >> fst;
    cin >> snd;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');
    return make_pair(fst, snd);
}

Emails readInput() {
    size_t n{};
    cin >> n;
    cin.ignore(numeric_limits<streamsize>::max(), '\n');

    Emails emails{};
    for (size_t i{0}; i < n; ++i) {
        pair<string, string> email = readPair<string>();
        emails.push_back(email);
    }

    return emails;
}


vector<string> filterHasGmail(Emails emails) {
    vector<string> gmailAddr{};

    regex gmail_pattern(R"([a-z]+(\.[a-z]+)?@gmail\.com$)");
    string firstName{}, emailAddr{};
    for (const auto& email : emails) {
        tie(firstName, emailAddr) = email;
        if (regex_match(emailAddr, gmail_pattern)) {
            gmailAddr.push_back(firstName);
        }
    }
    
    return gmailAddr;
}


int main() {
    // read | filter | show

    Emails emails{ readInput() };

    vector<string> gmailUsers = filterHasGmail(emails);
    sort(gmailUsers.begin(), gmailUsers.end());

    for (const string& firstName : gmailUsers) {
        cout << firstName << endl;
    }

    return 0;
}
